File: doc/code/cpp/objc.cpp
#include <iostream>
#include <memory>

#include <LIEF/MachO.hpp>
#include <LIEF/ObjC.hpp>

void check() {
  // lief-doc: check-start
  std::unique_ptr<LIEF::MachO::Binary> macho;

  std::unique_ptr<LIEF::objc::Metadata> metadata = macho->objc_metadata();

  if (metadata != nullptr) {
    std::cout << "Objective metadata found\n";
  }
  // lief-doc: check-end
}

void inspect() {
  // lief-doc: inspect-start
  std::unique_ptr<LIEF::MachO::Binary> bin;

  std::unique_ptr<LIEF::objc::Metadata> metadata = bin->objc_metadata();

  for (const LIEF::objc::Class& clazz : metadata->classes()) {
    std::cout << "name=" << clazz.name() << '\n';
    for (const LIEF::objc::Method& meth : clazz.methods()) {
      std::cout << "  method.name=" << meth.name() << '\n';
    }
  }

  std::cout << metadata->to_decl();
  // lief-doc: inspect-end
}

File: doc/code/rust/src/objc.rs
use std::process;

pub fn check(some_macho: &lief::macho::Binary) {
    // lief-doc: check-start
    let macho: &lief::macho::Binary = some_macho;

    if let Some(metadata) = macho.objc_metadata() {
        println!("Objective-C metadata found");
    }
    // lief-doc: check-end
}

pub fn inspect(some_macho: &lief::macho::Binary) {
    // lief-doc: inspect-start
    let macho: &lief::macho::Binary = some_macho;

    let Some(metadata) = macho.objc_metadata() else {
        process::exit(1);
    };

    for class in metadata.classes() {
        println!("name={}", class.name());
        for method in class.methods() {
            println!("  method.name={}", method.name());
        }
    }
    println!("{}", metadata.to_decl());
    // lief-doc: inspect-end
}

// lief-doc: classdump-start
fn classdump(macho: &lief::macho::Binary) {
    let metadata = macho.objc_metadata().expect("Missing Objective-C info");
    for class in metadata.classes() {
        println!("{}", class.to_decl());
    }
}
// lief-doc: classdump-end
